### PR TITLE
Sound-triggered traps

### PR DESCRIPTION
**Pull request by psuedolet**
Sound-triggered traps. Add traps to sound processing, add sound_threshold json key for traps, and add a test noise-triggered trap.

#### Summary
Features "Sound-triggered traps"

#### Purpose of change
Create the possibility to have things happen from sounds. One step towards the possibility of dormant zombies #65305

#### Describe the solution
The map now stores types of traps that have a sound trigger. When processing sounds, we iterate over these traps, similar to creatures. Sound trigger can be added to traps with the json key "sound_threshold", which defines the minimum volume of noise that will trigger the trap.

#### Describe alternatives you've considered
Add items to sound processing, so corpses can be used for dormant zombies. This seems like a potentially large performance hit and large amount of code work.
Pass the sound to the trap act function, so traps can do things like repeat the sound louder or only trigger for speech. This may be a potential future PR, but I wanted to get the basic functionality working.

#### Testing
Start a new game. Add the test trap. Yelling triggered the trap. Walking near it sometimes triggered the trap. The noise from the trap triggering sometimes triggered the trap.

#### Additional context
The new noise trigger is an addition to the standard enter-same-tile trigger. It does not remove the standard trigger.
If a triggered trap makes a noise it can trigger other noise-triggered traps, including itself!

### CODE REVIEW

File: src/sounds.cpp [hunk 1]
```diff
@@ -481,6 +482,16 @@ void sounds::process_sounds()
                 critter.hear_sound( source, vol, dist, this_centroid.provocative );
             }
         }
+        // Trigger sound-triggered traps
+        for( const trap *trapType : trap::get_sound_triggered_traps() ) {
+            for( tripoint tp : get_map().trap_locations( trapType->id ) ) {
+                const int dist = sound_distance( source, tp );
+                trap tr = get_map().tr_at( tp );
```
Comment (ehughsbaird) [inline src/sounds.cpp:489]: No need to copy these.
```suggestion
            for( const tripoint &tp : get_map().trap_locations( trapType->id ) ) {
                const int dist = sound_distance( source, tp );
                const trap &tr = get_map().tr_at( tp );
```

File: src/trap.h [hunk 2]
```diff
@@ -331,6 +338,13 @@ struct trap {
         static const std::vector<const trap *> &get_funnels();
         /*@}*/
 
+        /*
+         * Can the trap be triggered by sounds?
+         */
+        bool has_sound_trigger() const;
+        static const std::vector<const trap *> &get_sound_triggered_traps();
+        bool triggered_by_sound( const int vol, const int dist );
```
Comment (ehughsbaird) [inline src/trap.h:346]: As the CI asks
```suggestion
        bool triggered_by_sound( int vol, int dist );
```

**Comment by GuardianDll** [comment]
Please document your change in https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/JSON_INFO.md#traps

**Comment by ilovehans10** [comment]
> The noise from the trap triggering sometimes triggered the trap.

Couldn't this cause issues in some cases where a trap triggers itself indefinitely. I can see how that feature has some good uses, but would adding an optional cooldown time variable to traps allow for a cheap and easy way to avoid recursive traps.

**Comment by psuedolet** [comment]
> 



> > The noise from the trap triggering sometimes triggered the trap.
> 
> Couldn't this cause issues in some cases where a trap triggers itself indefinitely. I can see how that feature has some good uses, but would adding an optional cooldown time variable to traps allow for a cheap and easy way to avoid recursive traps.

That would only happen if 
1. The trap has both a sound trigger and an effect that produces a sound
2. The sound produced is at least as loud as the sound trigger threshold
3. The trap persists after triggering

**Pull request merged**